src/hotspot/share/gc/g1/g1ServiceThread.cpp

```diff
@@ -155,19 +149,259 @@ class G1YoungRemSetSamplingClosure : public HeapRegionClosure {
+  assert(_monitor.owned_by_self(), "Must be owner of lock");
+  assert(!_task_queue.is_empty(), "Should not be called for empty list");
+
+  double time_diff = _task_queue.peek()->time() - os::elapsedTime();
+  if (time_diff < 0) {
+    // Run without sleeping.
+    return 0;
+  }
+
+  // Return sleep time in milliseconds.
+  return (int64_t) (time_diff * MILLIUNITS);
+}
+
+void G1ServiceThread::sleep_before_next_cycle() {
+  if (should_terminate()) {
+    return;
+  }
+
+  MonitorLocker ml(&_monitor, Mutex::_no_safepoint_check_flag);
+  if (_task_queue.is_empty()) {
+    // Sleep until new task is registered if no tasks available.
+    log_trace(gc, task)("G1 Service Thread (wait for new tasks)");
+    ml.wait(0);
+  } else {
+    int64_t sleep_ms = time_to_next_task_ms();
+    if (sleep_ms > 0) {
+      log_trace(gc, task)("G1 Service Thread (wait) %1.3fs", sleep_ms / 1000.0);
+      ml.wait(sleep_ms);
+    }
+  }
+}
+
+void G1ServiceThread::reschedule_task(G1ServiceTask* task) {
+  if (!task->should_reschedule()) {
+    log_trace(gc, task)("G1 Service Thread (%s) (done)", task->name());
+    return;
+  }
+
+  // Reschedule task by updating task time and add back to queue.
+  double delay_ms = task->delay_ms() / 1000.0;
+  task->set_time(os::elapsedTime() + delay_ms);
+
+  MutexLocker ml(&_monitor, Mutex::_no_safepoint_check_flag);
+  _task_queue.add_ordered(task);
+
+  log_trace(gc, task)("G1 Service Thread (%s) (schedule) @%1.3fs", task->name(), task->time());;
+}
+
+G1ServiceTask* G1ServiceThread::pop_due_task() {
+  MutexLocker ml(&_monitor, Mutex::_no_safepoint_check_flag);
+  if (_task_queue.is_empty() || time_to_next_task_ms() != 0) {
+    return NULL;
+  }
+
+  return _task_queue.pop();
+}
+
+void G1ServiceThread::run_task(G1ServiceTask* task) {
+  double start = os::elapsedVTime();
+
+  log_debug(gc, task, start)("G1 Service Thread (%s) (run)", task->name());
+  task->execute();
+
+  double duration = os::elapsedVTime() - start;
+  log_debug(gc, task)("G1 Service Thread (%s) (run) %1.3fms", task->name(), duration * MILLIUNITS);
+}
+
+void G1ServiceThread::run_tasks() {
+  // Execute tasks that are due. Need to check for termination to avoid
+  // running forever if task reschedule with very short interval.
+  while (!should_terminate()) {
+    G1ServiceTask* task = pop_due_task();
+    if (task == NULL) {
+      return;
+    }
+
+    run_task(task);
+    reschedule_task(task);
+  }
+}
+
+void G1ServiceThread::run_service() {
+  double vtime_start = os::elapsedVTime();
+
+  // Setup the tasks handeled by the service thread and
+  // add them to the task list.
+  G1PeriodicGCTask gc_task("Periodic GC Task");
+  register_task(&gc_task);
+
+  G1RemSetSamplingTask remset_task("Remembered Set Sampling Task");
+  register_task(&remset_task);
+
+  while (!should_terminate()) {
+    run_tasks();
+
+    if (os::supports_vtime()) {
+      _vtime_accum = (os::elapsedVTime() - vtime_start);
+    } else {
+      _vtime_accum = 0.0;
```

Comment:
Why resetting it to zero? I thought empty else branch is enough here.

Reply:
I think you are correct but this is pre-existing code that I don't intend to change right now. There is a separate [issue](https://bugs.openjdk.java.net/browse/JDK-8252645) on moving this to the G1RemSetSamplingTask. Back in the day when the thread only did remembered set sampling it made sense to measure it here, it doesn't anymore, but I wanted to leave that out of this change.

Reply:
Got it, sure.

Reply:
Okay, let's postpone this to 8252645.